Commented/cleaned up code in patientSignUp html, css, and js files

Mannendri opened the pull request:
Fixed:
* Form now redirects to patientWaitingRoom.html whereas before it redirected to `/`

Cleanup
* Deleted an unused login form
* Deleted dead code (JS, html, css)
* Added comments

This partially addresses #50

Mannendri committed:
Commented/cleaned up code in patientSignUp.html, patientSignUp.css, and patientSignUp.js

j0n3lson commented on public/components/patientSignUp.js:
```diff
@@ -38,21 +38,6 @@ $('.tab a').on('click', function (e) {
-// submitForm.addEventListener("click", formSubmit);
-
 function formSubmit() {
   alert (firstName);
```
Probably delete this. We don't want a pop up.

j0n3lson commented on public/components/patientSignUp.js:
```diff
@@ -38,21 +38,6 @@ $('.tab a').on('click', function (e) {
-// }
-// submitForm.addEventListener("click", formSubmit);
-
 function formSubmit() {
```
Maybe we should call this validateInput()? And maybe we we should validate all form text fields. We should have the following checks. Since we're already using jQuery maybe we should implement the checks in jQuery:

**Required:**
* Field cannot be empty
* Field must contain at least one alpha numeric character
* Field cannot contain only spaces

Here's a good [stackoverflow](https://stackoverflow.com/a/24429337/478815) solution for form validation.

lillianchou94 replied:
1. We don't need to programmatically validate inputs (as in validate in code) if the input elements have "required" property set. I think in our case, the inputs have "required" on them. See https://www.w3schools.com/tags/att_input_required.asp

2. This function is called when the sign up submit [button](https://github.com/mellychat/webapp/blob/2aaf7972c53253ac4e711d17b64bac44f0ecefb9/public/templates/patientSignUp.html#L111) is clicked and all the inputs are validated via #1. So what is the goal of this function? Is it to redirect the user somewhere? Should that logic be in this function instead of [here](https://github.com/mellychat/webapp/blob/2aaf7972c53253ac4e711d17b64bac44f0ecefb9/public/components/patientSignUp.js#L52)?

j0n3lson replied:
Good point. I opened a bug about this but I was thinking about what we'll do with this data. So maybe validation might mean "check if the user is already existing in the database". So you're right that these simple validations are needed but I think some validation is required.

So to summarize maybe we do the following to cover the simple cases (on the html elements)
* Use the `maxlength` attribute
* Use the `required` attribute

Thoughts?

j0n3lson replied:
Resolving this thread. We'll move forward with eliminating this validation logic and using `maxLength` and `required` attributes on these form elements.

j0n3lson commented on public/components/patientSignUp.js:
```diff
@@ -38,21 +38,6 @@ $('.tab a').on('click', function (e) {
-
 function formSubmit() {
   alert (firstName);
   if (firstName.value == ""){
```
This sort of check can be improved by trimming spaces and then checking the length, see this [stackoverflow](https://stackoverflow.com/questions/154059/how-can-i-check-for-an-empty-undefined-null-string-in-javascript) entry.

j0n3lson replied:
See Lilly's reply below, we can eliminate this validation logic using `required` and `maxLength` attributes on the element:

https://github.com/mellychat/webapp/pull/47#discussion_r549049477

Let's remove this validation logic.

j0n3lson committed:
Removed unused submitForm variable

j0n3lson commented on public/components/patientSignUp.js:
```diff
@@ -38,21 +38,6 @@ $('.tab a').on('click', function (e) {
-
 function formSubmit() {
   alert (firstName);
   if (firstName.value == ""){
```
Note for the if-elseif-else here. You can eliminate the extra `else-if` clause when the previous clause has a return:

```
if(firstName.value == ""){  // Clause#1
  return false;
}
if(lastName.value == ""){ // Clause#2
  return false;
}
return true; // Final Clause
```

Notice how there aren't an `else` clauses in there. It's not needed since the previous clause (Clause#1) will return and end the method early.

**Feel free to resolve this comment.**

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,85 @@
+            <!-- ***** Sign Up Area Start ***** -->
+            <div>
+              <h1 class="formText">Fill The Form To Get Matched!</h1>
+              <form action="/" method="post">
```
Please give this element an id, say `id=signup-form`

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,85 @@
+                  <!-- Input field for first name -->
+                  <div class="field-wrap">
+                      <label>First Name<span class="req">*</span> </label>
+                    <input type="text" name="firstName" id="firstName"  autocomplete="off" required/>
```
Nit, could you put the `id` attribute before the `name` attribute. Make this consistent for all elements.

j0n3lson replied:
Let's set `maxlength=15` for both `firstName` and `lastName` fields. We don't want someone to put an entire book here.

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,85 @@
+                <div class="field-wrap">
+                  <textarea
+                    id="textArea"
+                    name="paragraph_text"
```
Can you make the `name` here consistent camel case.

Mannendri replied:
Based on the most recent convos, it seems like we are using the conventions found on Google Style Guide. I updated all the class/ID names to match this. Can I assume that this also applies to this name attribute? (I couldn't find anything about the name attribute on the style guide)
So instead of "paragraph_text", we use "paragraph-text"?

Mannendri replied:
I did this in my most recent commit, assuming the answer would be yes.

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,85 @@
+                <!-- ***** Top Row End ***** -->
+                <!-- Input field for description of issue -->
+                <div class="field-wrap">
+                  <textarea
```
Can you make this a required field as well.

j0n3lson replied:
Can you set `maxlength=300` so that the field doesn't have an unlimited character size.

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,85 @@
+            <!-- ***** Sign Up Area End ***** -->
+
+            <!-- ***** Login Area Start ***** -->
+            <!-- CONFUSION: THE BELOW CODE SEEMS TO IMPACT THE WAY THE ABOVE FORM IS RENDERED BUT I DON'T KNOW WHY THIS IS THE CASE WHEN THE BELOW CODE IS FOR A LOGIN FORM NOT USED ON THIS PAGE -->
```
I'd say let's delete it. We don't have support for logging in so this is a bit premature. When we get around to adding login capability, we can add it back.

For now, I'd say delete this code.

lillianchou94 replied:
Great point! It is super odd that this code block impacts the way the entire form is rendered. It is likely because the entire code for the form in patient sign up is copied over from practitioner sign up (see [prSignUp.css import in this file](https://github.com/mellychat/webapp/blob/2aaf7972c53253ac4e711d17b64bac44f0ecefb9/public/templates/patientSignUp.html#L19))

The way this form is formatted is mostly done in the css file. It's even more confusing if the same selectors are being imported in prSignUp.css and patientSignUp.css. If you have two css rules present, the last one will take precedence. Maybe you could start the investigation there if you want to delete the signup block safely without breaking the entire form :)

Mannendri replied:
Okay, so I was able to resolve this. There was a CSS selector in prSignUp.css that applied a "display:none" on the last  child of the div ".tab-content". When I deleted this login code, what was happening was that the sign up form was the last child and ended up being hidden. So what I did was change the CSS rule to do a display:none specifically on divs with an id of login: 

.tab-content  > #login{
display:none
}

I deleted the login form entirely from patientSignUp.html because it's unnecessary, but this style ^ is helpful for the other file that is linked to prSignUp.css (i.e. prSignUp.html).

Side Note: Apparently, there are some essential styles from prSignUp.css for patientSignUp.html to render properly so I can't eliminate the link to this stylesheet in that html file.

j0n3lson reviewed (changes requested):
Good job finding this other bugs -- I opened issues for tracking the bugs you found:

* #49 Nav bar
* #50 Redirection needs more design

For now, let's address these comments and focus on getting basic redirection working.

lillianchou94 commented on public/components/patientSignUp.js:
```diff
@@ -66,45 +50,6 @@ function formSubmit() {
 }
 
 if(formSubmit()) {
```
Some things to consider when trying to solve the problem you brought up:  _when_ this if statement is actually executed? Is it when we first load the page or is it when the form is submitted?

Mannendri replied:
I moved the redirection code into the function formSubmit which gets called when the submit button is pressed.

lillianchou94 commented on public/templates/patientSignUp.html:
```diff
@@ -45,12 +49,13 @@
                         <!-- ***** Logo End ***** -->
                         <!-- ***** Menu Start ***** -->
-                        <ul class="nav">
+                        <ul class="nav"> <!-- PROBLEM: THE NAV BAR APPEARS INVISIBLE WHEN APP IS RUN -->
```
It looks like the hamburger menu shows on my screen first, clicking on it exposes these nav items.

When I expand my screen, the hamburger icon goes away and all the nav items are actually on the page. But you can't see them because the font color is white. I think the background of the page is supposed to be blue so the white fonts would show. Maybe that's the issue here?

j0n3lson replied:
Can we address that in #30 so this PR can make progress?

Mannendri replied:
Okay, I will move the convo over there.

lillianchou94 commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,85 @@
     <div class="form">
-      <div class="signup">
+          <!-- ***** Patient Topic Selection Start ***** -->
           <h1 class="formText">Which topic relates to your issue?</h1>
```
If you're up for it, I'd update all the css selectors that uses camel casing to use hyphens

https://google.github.io/styleguide/htmlcssguide.html#ID_and_Class_Name_Delimiters

j0n3lson replied:
+1 I was looking into ways to use actions to lint the project so we can have a consistent standard.

For now we can just adopt that style guide I think :)

j0n3lson replied:
I opened #56 for this one Lilly. Seems like a good first issue to work on in a follow-up PR.

Closing out this thread and we'll address in that issue.

lillianchou94 reviewed (commented):
Good job!!!

Mannendri commented:
I am almost done addressing all of these issues. I will probably commit tomorrow and resolve each of these convos with an explanation (where appropriate).

j0n3lson commented:
> I am almost done addressing all of these issues. I will probably commit tomorrow and resolve each of these convos with an explanation (where appropriate).

Awesome! Great work on this PR (I know there were a lot of comments).

Let us know if you need help!

j0n3lson commented:
Friendly ping. Either a review is pending or changes are requested

Mannendri committed:
Updated all class/ids to match Google Style Guide; eliminated unnceessary validation logic in patientSignUp.js; removed unnecessary login HTML code safely

Mannendri commented:
Okay, so I commited resolving everything except for the basic redirection to patWaitingRoom.html and the nav bar visibility issue (still working on these). Hopefully I didn't overlook anything, but here's a general overview of what I did:
- Updated all classes/id to match Google Style Guide
- Removed unnecessary validation logic in patientSignUp.js (adding the maxlength and required attributes to the html elements)
- Safely eliminated unnecessary login form html code in patientSignUp.html

Mannendri commented:
As I mentioned [here](https://github.com/mellychat/webapp/issues/50#issuecomment-753229551) I can't figure out what's going on with basic redirection.

j0n3lson commented on public/assets/css/prSignUp.css:
```diff
@@ -98,7 +98,7 @@ a:hover {
   color: #fff;
 }
-.tab-content > div:last-child {
+.tab-content > #login {
```
Do we need this now that we've removed the login form?

Mannendri replied:
Yes, I believe we do still need this. We only removed the login form from patientSignUp.html, but prSignUp.html contains both a sign up form and a login form (displayed as two tabs on the webpage). This style prevents the login form from appearing below the sign up form when prSignUp is rendered.

Mannendri replied:
The original design intent was that the practitioner has the ability to both sign up and login, while the patient only has a sign up (a query regarding their health).

j0n3lson replied:
Ahh gotcha. Thanks for the clarification!

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -45,12 +49,13 @@
                         <!-- ***** Logo End ***** -->
                         <!-- ***** Menu Start ***** -->
-                        <ul class="nav">
+                        <ul class="nav"> <!-- PROBLEM: THE NAV BAR APPEARS INVISIBLE WHEN APP IS RUN - I am still working on this -->
```
Rather than a comment, just reference an existing issue (or open a new one if one doesn't exists):
```
// TODO(https://github.com/mellychat/webapp/issues/30): Nav bar is invisible when run
```

You don't need to fix this in this PR however.

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,64 @@
+            <!-- ***** Sign Up Area Start ***** -->
+            <div>
+              <h1 class="form-text">Fill The Form To Get Matched!</h1>
+              <form id="sign-up-form" action="/" method="post">
```
Setting the `action` attribute says send the form data to given URL (here the web root). So I think there's a few options to get basic redirection working: 

**Option 1 (Preferred):** Use `action=/templates/patWaitingRoom.html` and remove the `onclick` attribute from the button. Because the button has `type=submit` it will submit the form it's in.

**Option 2:** Keep using the `onclick` on the button but use the `window.location.assign()` method to change the window location inside the `formSubmit` method. Note that you can technically just have `onclick=window.location.assign(...)` directly on the button rather than having in javascript.

It seems like **Option 1** is the the right thing to do. Option 2 doesn't work because it triggers redirect even if the form fields are empty making `required` and `maxlength` are useless (tested this locally on Chrome).

Mannendri replied:
I implemented Option 1 and it works! Thank you

j0n3lson commented on public/templates/patientSignUp.html:
```diff
@@ -61,92 +66,64 @@
+    <!-- ***** Form End ***** -->
+
+    <!-- jQuery -->
     <script src="/assets/js/jquery-2.1.0.min.js"></script>
```
Can you move the script loads into the `head` section? It seems like they are down here so they load after the page has loaded -- you can just add a `defer` attribute to the `script` element if we really want to defer loading these scripts.

Mannendri replied:
Done.

Mannendri replied:
@j0n3lson , should we standardize this across the codebase?

j0n3lson replied:
Opened #60 for this standardization.

j0n3lson reviewed (changes requested):
Almost there. Great follow through @Mannendri

Mannendri committed:
Got basic redirection to patWaitingRoom.html working by setting action attribute; moved script loads into head section and added -defer- attributes

j0n3lson commented:
Hey @Mannendri 

Hey I edited this PR directly so you might get an error message when trying to push. If you get an error message like [this stackoverflow question](https://stackoverflow.com/questions/24114676/git-error-failed-to-push-some-refs-to-remote) post complaining about:

> Updates were rejected because the tip of your branch is behind ...

Just go ahead and force push:

> git push upstream Code-documentation-on-patientSignUp --force

It will overwrite my changes (which aren't needed) and allow you to push.

Mannendri commented:
I was able to push without having to do a force push. So we should be able to merge this PR now, right?

j0n3lson commented:
> I was able to push without having to do a force push. So we should be able to merge this PR now, right?

I wouldn't merge as it is since this PR will contain a whole bunch of unrelated commits.

Seems like your commit history is off. There's 21 commits in this PR. They are all the commits that happened after this PR was created. I'd expect only a few commits here.

Did you rebase correctly? I think you might have did a `git pull` instead of the normal rebase workflow?

Mannendri commented:
I did do the rebase correctly. But after rebasing, I believe I did a git pull (I'm not sure why; I shouldn't have). Would rebasing again fix this issue?

Mannendri commented:
Rebasing would not fix this issue, would I have to revert the branch to an earlier state?

j0n3lson commented:
> But after rebasing, I believe I did a git pull (I'm not sure why; I shouldn't have). 

You almost never want to do a `git pull` because that does a `git fetch` followed by a `git merge`. The merge says "take the commits from upstream and merge them into the commits in my local branch"

But when we do `git rebase` your saying "put the commits in my local branch on top of the last commit from upstream".


> Would rebasing again fix this issue?

I don't think a straight rebase will work (based on what rebase does). I think you need to do some surgery on your commit history. I wouldn't recommend doing this yourself -- I'll help you through it when we meet. Just putting this here for my notes:

1. Checkout a new branch based on this one
1. `git rebase -i` to interactively rebase
1. Drop any commits that aren't related to this PR (they are already merged so it's okay). This gets us back to the the state before you did `git pull`
1. Then do a proper rebase: `git rebase mainline`

Mannendri commented:
Got it. In the future, I will avoid doing a `git pull` (I apologize for not being more careful about this). For now, I guess we will postpone work on this PR to Friday. 

Thank you!

j0n3lson commented:
> Got it. In the future, I will avoid doing a `git pull` (I apologize for not being more careful about this). For now, I guess we will postpone work on this PR to Friday. 
> 
> Thank you!

Hey no worries. No need to apologise. Its a learning process ... Making mistakes is the only way to learn!☝🏿

Great work on this PR by the way. Love the follow through!

j0n3lson commented:
(pr-nagger-bot): Friendly ping. If you are a reviewer please take a look. If you are an author, please respond to comments.
